### DISCUSSION TITLE
Troubleshooting Start-Job

### DISCUSSION THREAD

**Discussion by markcandelora** [PowerShell/PowerShell#24382] [Q&A] [answered]
I'm trying to start a job, but keep getting access denied errors.  I'm running `pwsh` in a Linux Docker Dev Container, and can't even get a simple command to run.  There's definitely something wrong with permissions, but how do I troubleshoot this?

```
> Start-job -ScriptBlock { Write-Host 'hello' } | Receive-Job
OpenError: [localhost] An error occurred while starting the background process. Error reported: An error occurred trying to start process '/usr/share/powershell/.store/powershell.linux.x64/7.4.4/powershell.linux.x64/7.4.4/tools/net8.0/any/pwsh' with working directory '/workspaces/project'. Permission denied.
```

devcontainer.json:
```
{
	"name": "C# (.NET)",
	"build": {
		"dockerfile": "Dockerfile"
	},
	"features": {
		"ghcr.io/devcontainers/features/github-cli:1": {},
		"ghcr.io/devcontainers/features/docker-in-docker:2": {},
		"ghcr.io/devcontainers/features/powershell:1": {},
		"ghcr.io/meaningful-ooo/devcontainer-features/homebrew:2": {}
	},
	"customizations": {
		"vscode": {
			"extensions": [
				"ms-vscode.powershell",
				"formulahendry.dotnet-test-explorer",
				"DotJoshJohnson.xml",
				"k--kato.docomment",
				"adrianwilczynski.user-secrets"
			],
			"settings": {
				"[csharp]": {
					"editor.maxTokenizationLineLength": 2500,
					"editor.formatOnSave": true,
					"editor.formatOnSaveMode": "file"
				}
			}
		}
	},
	"postCreateCommand": {
		"Pester": "sudo pwsh -Command 'Install-Module Pester -Force -Scope AllUsers'",
		"Act": "brew install act",
		"GitVersion": "dotnet tool install --global GitVersion.Tool --version 6.0.2"
	}
}
```

Dockerfile:
```
FROM mcr.microsoft.com/devcontainers/dotnet:8.0
```

The PowerShell feature appears to be installed through here:
https://github.com/devcontainers/features/tree/main/src/powershell

In hindsight, maybe this is a better discussion for the devcontainers team?

**Comment by rhubarb-geek-nz** [comment]
Unfortunately "pwsh in a Linux container" gives very little context. Is this a docker? Did you create it? How is PowerShell installed in the container? What architecture is it?

From the error it reminds me of this problem https://github.com/PowerShell/PowerShell/issues/24090

Where powershell was installed as a dotnet tool and the PATH was affecting for running child processes.

If you have installed PowerShell as a dotnet tool, I recommend installing PowerShell from either a Deb/Rpm package or else from a tar.gz.

**Reply by markcandelora** [reply]
Good points, thank you.  Probably best not to start discussions at midnight after pulling my hair out!  I've updated the original post with more of the details on PowerShell installation.

I have rebuilt the container several times to no avail.  I'll dig deeper into the post you referenced above to see if there's any brain nuggets I can gleam.

**Comment by kasini3000** [comment]
I guess this is a deliberate restriction by PowerShell itself.

------

#### Can nginx running in Docker fork itself and become a child process?
ai :
Whether Nginx running in Docker containers can fork itself as a child process involves Docker's isolation mechanism and Nginx's running mode. Let's analyze step by step:

Process isolation in Docker:
Docker containers are essentially an encapsulation of a single process. By default, a container should run one main process.
Docker uses Linux's namespace technology to isolate processes, which means that although processes within a container can see a complete process tree, they are actually all in the same namespace.
Nginx's working mode:
Nginx typically uses a multiprocessing model to handle requests. The Master Process is responsible for managing and monitoring Worker Processes.
The main process can fork multiple worker processes to handle concurrent connections.
The actual situation in Docker:
Default configuration: If Nginx runs in a Docker container by default (i.e. without any special configuration to change its behavior), Nginx will indeed attempt to fork out worker processes. This behavior of Nginx will not change with Docker, as from Nginx's perspective, it still runs on a 'system'.
Limitations and Configuration:
Docker can limit certain capabilities of processes within a container through configuration, such as using -- pids limit to restrict the number of processes that can be forked, but this is not the default behavior.
If you use the -- pid=host parameter in Docker run, Nginx will be able to see the PID namespace of the host, but this will weaken isolation.
In some container orchestration tools such as Kubernetes, resource restrictions and security context can be set, which may affect Nginx's fork behavior.

Considerations in practical applications:
In a container environment, it is generally recommended to use a one container one process model to maintain simplicity and manageability of the environment. But for Nginx, running multiple processes (one main process and multiple worker processes) is its standard operating mode.
If it is necessary to strictly limit Nginx to run only one process, you can modify Nginx configuration (such as setting workr_processes 1;) or use Docker's entrypoint script to control Nginx's startup mode so that it does not fork child processes.
Summary:

Sure: By default, Nginx can fork child processes in Docker containers because this behavior is part of Nginx's design and Docker's design allows for this operation.

**Reply by rhubarb-geek-nz** [reply]
> I guess this is a deliberate restriction by PowerShell itself.

I suspect the limitation is caused by using the dotnet tool version of PowerShell with its PATHing problems. I use a standard `deb` packaging version of PowerShell in a docker image and am happily able to run jobs.

```
PS> Start-Job { 'foo' } | Receive-Job -Wait
foo
```

**Comment by markcandelora** [comment]
As a follow-up here, is there any way to get additional debugging information to figure out what operation is getting the access denied error? Common Parameters `-Debug` and `-Verbose` don't provide any additional output.

**Comment by rhubarb-geek-nz** [comment]
I suggest revisit the original error

```
OpenError: [localhost] An error occurred while starting the background process. Error reported: An error occurred trying to start process '/usr/share/powershell/.store/powershell.linux.x64/7.4.4/powershell.linux.x64/7.4.4/tools/net8.0/any/pwsh' with working directory '/workspaces/project'. Permission denied.
```

Log into the docker container with `bash` and look at the permissions on

- /workspaces/project
- /usr/share/powershell/.store/powershell.linux.x64/7.4.4/powershell.linux.x64/7.4.4/tools/net8.0/any/pwsh

and examine with `file`. 

What architecture are you running? You do not say.

Can you run that `pwsh` directly from `bash`?

Can you use `ldd` to examine the executable?

**Reply by markcandelora** [reply]
That just got to the heart of the issue (seems like you may have already knew that) - my current user doesn't have execute access to `pwsh`:

```
$ ls -l /usr/share/powershell/.store/powershell.linux.x64/7.4.4/powershell.linux.x64/7.4.4/tools/net8.0/any/pwsh
-rwxr--r-- 1 root root 72520 Jul 17 19:26 /usr/share/powershell/.store/powershell.linux.x64/7.4.4/powershell.linux.x64/7.4.4/tools/net8.0/any/pwsh

$ file /usr/share/powershell/.store/powershell.linux.x64/7.4.4/powershell.linux.x64/7.4.4/tools/net8.0/any/pwsh
/usr/share/powershell/.store/powershell.linux.x64/7.4.4/powershell.linux.x64/7.4.4/tools/net8.0/any/pwsh: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=1d61f38db3a6fee79644617383216ddef21b6e59, stripped
```

When I first saw this I didn't understand how this could be since I can start a PowerShell terminal just fine from the IDE, but in looking at this further I noticed that the path being used to invoke the shell from the IDE is `/usr/bin/pwsh`:

![image](https://github.com/user-attachments/assets/23c1afc7-1920-4aab-b5b6-5db19c528827)

I started investigating the PATH variable and found that pwsh adds `/usr/share/powershell/.store/powershell.linux.x64/7.4.4/powershell.linux.x64/7.4.4/tools/net8.0/any` to the path:

```
$ pwsh
PowerShell 7.4.4

   A new PowerShell stable release is available: v7.4.5 
   Upgrade now, or check out the release page at:       
     https://aka.ms/PowerShell-Release?tag=v7.4.5       

PS> $env:PATH
/usr/share/powershell/.store/powershell.linux.x64/7.4.4/powershell.linux.x64/7.4.4/tools/net8.0/any:/vscode/vscode-server/bin/linux-x64/38c31bc77e0dd6ae88a4e9cc93428cc27a56ba40/bin/remote-cli:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/usr/local/share/nvm/current/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/usr/local/share/nvm/current/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/vscode/.dotnet:/home/vscode/.dotnet/tools:/home/vscode/.local/bin:/home/vscode/.dotnet:/home/vscode/.dotnet/tools
```

**Comment by rhubarb-geek-nz** [comment, marked as answer]
I think that the powershell dotnet tool as a general purpose shell is a flawed concept, It works well executing small scripts within building a dotnet csproj, but it is too complex for its own good to work well as a general shell.

I recommend not using the tool and running PowerShell from a deb package or simply untarring into a directory ( /opt/microsoft/powershell/7 is a good place!)

**Reply by markcandelora** [reply]
I couldn't figure out initially why you were saying to avoid the dotnet tool install, but in another facepalm moment I realized that the pwsh path made that pretty clear.

I've done a bunch more digging and found that despite my using the PowerShell devcontainer feature (which does *not* use the dotnet tool install method), the [base image](https://github.com/dotnet/dotnet-docker/blob/main/src/sdk/8.0/bookworm-slim/amd64/Dockerfile) I was using already had PowerShell installed via dotnet tool (facepalm).

I'm working on a fix for it now (rolling my own image using the base aspnet image).  

Thanks again for the help, I'll post back with the final results of my effort for posterity :-D

**Reply by markcandelora** [reply]
For anyone using devcontainers with dotnet SDK and PowerShell will want to avoid using the devcontainer/dotnet image if you need PowerShell to be really functional.  My final solution was to use devcontainer features to install the dotnet SDK and PowerShell:

devcontainer.json:
```
{
	"name": "C# (.NET)",
	"build": {
		"dockerfile": "Dockerfile"
	},
	"features": {
		"ghcr.io/devcontainers/features/dotnet:2": { "version": "8.0" },
		"ghcr.io/devcontainers/features/powershell:1": {},
	},
	"remoteUser": "vscode"
```

dockerfile:
```
ARG VARIANT=8.0.8-bookworm-slim-amd64
FROM mcr.microsoft.com/dotnet/aspnet:${VARIANT}

ARG VARIANT
RUN if id "ubuntu" &>/dev/null; then \
        echo "Deleting user 'ubuntu'  for $VARIANT" && userdel -f -r ubuntu || echo "Failed to delete ubuntu user for $VARIANT"; \
    else \
        echo "User 'ubuntu' does not exist for $VARIANT"; \ 
    fi
```
